README.md
# RenderPlayerMod-1.8.9
[Modのダウンロード](https://github.com/S4wa/RenderPlayerMod-1.8.9/releases) / [ベース1](https://github.com/SimplyRin/RenderPlayerMod-1.8.9) / [ベース2](https://youtu.be/9jucA23WRtg)  
※バージョンを更新して様子がおかしい時はコンフィグファイルを消してください (configって名前のフォルダにある"renderplayermod.txt"です)。  

変更点  
・/renderplayer [right,left] から、/renderplayerに変え CPSModのような操作方法で好きな場所に設定できる。  
・ベースではコンフィグを読み込んだりするコードが書かれてなかったから書き足しました。  
・表示のオンオフができる。  


おかしな点   
(修正済み)~~・[ここ](https://goo.gl/zjg8JV)の変数のせいで位置変更する時に足にしかクリックの判定がなく、うまくドラッグできない。~~  
・位置変更の時にプレイヤーより手前に[ここ](https://goo.gl/tttvG2)が表示される。  
(修正済み)~~・HP以外のゲージのテクスチャが白くなっている。~~  


[![YouTube](https://i.imgur.com/QJZFRcK.jpg)](https://youtu.be/SWIBdEqSjts)


resources/changelog.txt
+------------------------------------------------------------------------------+
|                           v1.0.0 - 03/08/2018                                |
+------------------------------------------------------------------------------+
- HP以外のゲージが白く潰れるバグを消滅させた。
- 普段のプレイに支障が出る不具合はなくなったのでv1.0.0リリース。




+------------------------------------------------------------------------------+
|                           v0.0.5 - 03/08/2018                                |
+------------------------------------------------------------------------------+
- 頭の向きをスライダーで設定できるようにした。
- サイズを変更するスライダーのリセットボタンをスライダーの上から横に移動。




+------------------------------------------------------------------------------+
|                           v0.0.4 - 03/07/2018                                |
+------------------------------------------------------------------------------+
- サイズをスライダーで設定できるようにした。




+------------------------------------------------------------------------------+
|                           v0.0.3 - 03/07/2018                                |
+------------------------------------------------------------------------------+
- "reload config"の追加と、位置変更する時のクリックの判定の修正。




+------------------------------------------------------------------------------+
|                           v0.0.2 - 03/07/2018                                |
+------------------------------------------------------------------------------+
- 首の向きが正面に向くよう修正。




+------------------------------------------------------------------------------+
|                           v0.0.1 - 03/07/2018                                |
+------------------------------------------------------------------------------+
- 表示のオンオフができるようになった。

net/simplyrin/renderplayermod/GuiSliderFixed.java
package net.simplyrin.renderplayermod;

import org.lwjgl.opengl.GL11;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.gui.GuiButton;
import net.minecraft.client.renderer.GlStateManager;

public class GuiSliderFixed extends GuiButton {
	public float sliderValue = 1.0F;
	public float sliderMaxValue = 1.0F;
	public float sliderMinValue = 1.0F;
	public boolean dragging = false;
	public String label;

	public GuiSliderFixed(int id, int x, int y, String label, float startingValue, float maxValue, float minValue) {
		super(id, x, y, 150, 20, label);
		this.label = label;
		this.sliderValue = startingValue;
		this.sliderMaxValue = maxValue;
		this.sliderMinValue = minValue;
	}

	protected int getHoverState(boolean par1) {
		return 0;
	}

	@Override

	public void drawButton(Minecraft mc, int mouseX, int mouseY) {
		if (this.visible) {
			FontRenderer fontrenderer = mc.fontRendererObj;
			mc.getTextureManager().bindTexture(buttonTextures);
			GlStateManager.color(1.0F, 1.0F, 1.0F, 1.0F);
			this.hovered = mouseX >= this.xPosition && mouseY >= this.yPosition &&
					mouseX < this.xPosition + this.width && mouseY < this.yPosition + this.height;
			int k = this.getHoverState(this.hovered);
			GlStateManager.enableBlend();
			GlStateManager.tryBlendFuncSeparate(770, 771, 1, 0);
			GlStateManager.blendFunc(770, 771);

			this.drawTexturedModalRect(this.xPosition, this.yPosition, 0, 46 + k * 20, this.width / 2, this.height);
			this.drawTexturedModalRect(this.xPosition + this.width / 2, this.yPosition, 200 - this.width / 2,
					46 + k * 20, this.width / 2, this.height);
			this.mouseDragged(mc, mouseX, mouseY);
			int l = 14737632;
			if (packedFGColour != 0) {
				l = packedFGColour;
			} else if (!this.enabled) {
				l = 10526880;
			} else if (this.hovered) {
				l = 16777120;
			}
			this.drawCenteredString(fontrenderer, this.displayString, this.xPosition + this.width / 2,
					this.yPosition + (this.height - 8) / 2, l);
		}
	}

	protected void mouseDragged(Minecraft par1Minecraft, int par2, int par3) {
		if (this.enabled && this.visible && this.packedFGColour == 0) {
			if (this.dragging) {
				this.sliderValue = (float) (par2 - (this.xPosition + 4)) / (float) (this.width - 8);
				if (this.sliderValue < 0.0F) {
					this.sliderValue = 0.0F;
				}
				if (this.sliderValue > 1.0F) {
					this.sliderValue = 1.0F;
				}
			}
			this.displayString = label + ": " + (int) (sliderValue * sliderMaxValue);
			GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
			this.drawTexturedModalRect(this.xPosition + (int) (this.sliderValue * (float) (this.width - 8)),
					this.yPosition, 0, 66, 4, 20);
			this.drawTexturedModalRect(this.xPosition + (int) (this.sliderValue * (float) (this.width - 8)) + 4,
					this.yPosition, 196, 66, 4, 20);
		}
	}

	public boolean mousePressed(Minecraft par1Minecraft, int par2, int par3) {
		if (super.mousePressed(par1Minecraft, par2, par3)) {
			this.sliderValue = (float) (par2 - (this.xPosition + 4)) / (float) (this.width - 8);
			if (this.sliderValue < 0.0F) {
				this.sliderValue = 0.0F;
			}
			if (this.sliderValue > 1.0F) {
				this.sliderValue = 1.0F;
			}
			this.dragging = true;
			return true;
		} else {
			return false;
		}
	}

	public void mouseReleased(int par1, int par2) {
		this.dragging = false;
	}
}

net/simplyrin/renderplayermod/RenderPlayerMod.java
package net.simplyrin.renderplayermod;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.inventory.GuiInventory;
import net.minecraftforge.client.ClientCommandHandler;
import net.minecraftforge.client.event.RenderGameOverlayEvent;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.Mod.EventHandler;
import net.minecraftforge.fml.common.event.FMLInitializationEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import net.simplyrin.renderplayermod.command.RenderPlayer;

@SideOnly(Side.CLIENT)
@Mod(modid = RenderPlayerMod.MODID, version = RenderPlayerMod.VERSION, clientSideOnly = true)
public class RenderPlayerMod {
	public static final String MODID = "RenderPlayerMod";
	public static final String VERSION = "0.0.5 PRE-RELEASE";
	private final Minecraft mc;

	public RenderPlayerMod() {
		this.mc = Minecraft.getMinecraft();
	}

	@EventHandler
	public void init(FMLInitializationEvent event) {
		MinecraftForge.EVENT_BUS.register(this);
		ClientCommandHandler.instance.registerCommand(new RenderPlayer());

		File file = new File("config/renderplayermod.txt");
		loadSettings();
	}

	@SubscribeEvent
	public void onRenderGui(RenderGameOverlayEvent.Post event) {
		if ((!RenderPlayerMod.enabled) || (this.mc.gameSettings.showDebugInfo)) return;
		GuiInventory.drawEntityOnScreen(counterPosX, counterPosY, size, horizontally, vertically, mc.thePlayer);
	}

	public static String getPrefix() {
		return "§7[§cRenderPlayer§7] §r";
	}

	private static void loadSettings() {
		File settings = new File("config/renderplayermod.txt");
		if (!settings.exists()) {
			try {
				settings.createNewFile();
				BufferedWriter writer = new BufferedWriter(new FileWriter(settings));
				writer.write(counterPosX + ":" + counterPosY + ":" + enabled + ":" + size + ":" + horizontally + ":" + vertically);
				writer.close();
			} catch (FileNotFoundException err) {
				err.printStackTrace();
			} catch (IOException err) {
				err.printStackTrace();
			}
			return;
		} else {
			try {
				BufferedReader reader = new BufferedReader(new FileReader(settings));
				String[] options = reader.readLine().split(":");
				counterPosX = Integer.valueOf(options[0]).intValue();
				counterPosY = Integer.valueOf(options[1]).intValue();
				enabled = StringToBoolean(options[2]);
				size = Integer.valueOf(options[3]).intValue();
				horizontally = Float.parseFloat(options[4]);
				vertically = Float.parseFloat(options[5]);
				reader.close();
			} catch (FileNotFoundException e) {
				e.printStackTrace();
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	public static void saveSettings() {
		File settings = new File("config/renderplayermod.txt");
		try {
			BufferedWriter writer = new BufferedWriter(new FileWriter(settings));
			writer.write(counterPosX + ":" + counterPosY + ":" + enabled + ":" + size + ":" + horizontally + ":" + vertically);
			writer.close();
		} catch (IOException err) {
			err.printStackTrace();
		}
	}

	public static void reloadSettings() {
		File settings = new File("config/renderplayermod.txt");
		if (!settings.exists()) return;
		try {
			BufferedReader reader = new BufferedReader(new FileReader(settings));
			String[] options = reader.readLine().split(":");
			counterPosX = Integer.valueOf(options[0]).intValue();
			counterPosY = Integer.valueOf(options[1]).intValue();
			enabled = StringToBoolean(options[2]);
			size = Integer.valueOf(options[3]).intValue();
			horizontally = Float.parseFloat(options[4]);
			vertically = Float.parseFloat(options[5]);
			reader.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	private static boolean StringToBoolean(String arg) {
		arg = arg.toLowerCase();
		if ("true".equals(arg) || "1".equals(arg)) {
			return (true);
		} else if ("false".equals(arg) || "0".equals(arg)) {
			return (false);
		} else {
			throw (new IllegalArgumentException(arg));
		}
	}

	public static boolean enabled = true;
	public static boolean guisettings = false;
	public static int counterPosX = 20;
	public static int counterPosY = 60;
	public static int size = 30;
	public static float horizontally = 0;
	public static float vertically = 0;
}


net/simplyrin/renderplayermod/utilities/GuiSlideControl.java
package net.simplyrin.renderplayermod.utilities;

import java.text.DecimalFormat;

import org.lwjgl.opengl.GL11;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiButton;

public class GuiSlideControl extends GuiButton {
	public String	label;			// Text displayed next to value
	public float	curValue;		// Current Slider position, from 0 to 1
	public float	minValue;		// Smallest value represented by the Slider
	public float	maxValue;		// Largest value represented by the Slider
	public boolean	isSliding;		// Is Control currently being manipulated?
	public boolean	useIntegers;	// Use Integers to display values

	private static DecimalFormat numFormat = new DecimalFormat("#.00");

	public GuiSlideControl(int id, int x, int y, int width, int height, String displayString, float minVal,
			float maxVal, float curVal, boolean useInts) {
		super(id, x, y, width, height,
				(useInts ? displayString + ((int) curVal) : displayString + numFormat.format(curVal)));

		this.label = displayString;
		this.minValue = minVal;
		this.maxValue = maxVal;
		this.curValue = (curVal - minVal) / (maxVal - minVal);
		this.useIntegers = useInts;
	}

	public float GetValueAsFloat() {
		return (maxValue - minValue) * curValue + minValue;
	}

	public int GetValueAsInt() {
		return (int) ((maxValue - minValue) * curValue + minValue);
	}

	protected float roundValue(float value) {
		value = 0.01F * (float) Math.round(value / 0.01F);
		return value;
	}

	public String GetLabel() {
		if (useIntegers)
			return label + GetValueAsInt();
		else
			return label + numFormat.format(GetValueAsFloat());
	}

	protected void SetLabel() {
		displayString = GetLabel();
	}

	@Override
	protected int getHoverState(boolean isMouseOver) {
		return 0;
	}

	@Override
	protected void mouseDragged(Minecraft mc, int mousePosX, int mousePosY) {
		if (visible) {
			if (this.isSliding) {
				curValue = roundValue((float) (mousePosX - (xPosition + 4)) / (float) (width - 8));

				if (curValue < 0.0F) {
					curValue = 0.0F;
				}
				if (curValue > 1.0F) {
					curValue = 1.0F;
				}

				SetLabel();
			}

			GL11.glColor4f(1.0F, 1.0F, 1.0F, 1.0F);
			this.drawTexturedModalRect(xPosition + (int) (curValue * (float) (width - 8)), yPosition, 0, 66, 4, 20);
			this.drawTexturedModalRect(xPosition + (int) (curValue * (float) (width - 8)) + 4, yPosition, 196, 66, 4,
					20);
		}
	}

	@Override
	public boolean mousePressed(Minecraft mc, int mousePosX, int mousePosY) {
		if (super.mousePressed(mc, mousePosX, mousePosY)) {
			curValue = roundValue((float) (mousePosX - (xPosition + 4)) / (float) (width - 8));

			if (curValue < 0.0F) {
				curValue = 0.0F;
			}
			if (curValue > 1.0F) {
				curValue = 1.0F;
			}

			SetLabel();
			this.isSliding = true;
			return true;
		} else {
			return false;
		}
	}

	@Override
	public void mouseReleased(int mousePosX, int mousePosY) {
		this.isSliding = false;
	}
}


net/simplyrin/renderplayermod/GuiSettings.java
package net.simplyrin.renderplayermod.command;

import java.awt.Desktop;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;

import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiButton;
import net.minecraft.client.gui.GuiScreen;
import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;
import net.simplyrin.renderplayermod.RenderPlayerMod;
import net.simplyrin.renderplayermod.utilities.GuiSlideControl;

public class GuiSettings extends GuiScreen {
	private boolean isDragging;
	private int lastX;
	private int lastY;
	private GuiButton buttonToggle;
	private GuiButton buttonReset;
	private GuiButton buttonResetSize;
	private GuiButton buttonReload;
	private GuiSlideControl sliderSize;
	private GuiSlideControl sliderHorizontally;
	private GuiSlideControl sliderVertically;
	private GuiButton buttonHorizontallyReset;
	private GuiButton buttonVerticallyReset;
	private GuiButton buttonOpenGithub;

	public GuiSettings() {
		this.isDragging = false;
		this.lastX = 0;
		this.lastY = 0;
	}

	public void initGui() {
		drawCenteredString(this.mc.fontRendererObj, "RenderPlayerMod 1.8.9", this.width / 2, 4, -1);
		this.buttonList.add(this.buttonToggle = new GuiButton(0, this.width / 2 - 75, this.height / 2 - 44, 150, 20,
				"Enabled: " + RenderPlayerMod.enabled));
		this.buttonList.add(this.buttonReset = new GuiButton(1, this.width / 2 - 75, this.height / 2 - 22, 150, 20,
				"Reset Position"));
		this.buttonList.add(this.buttonReload = new GuiButton(2, this.width / 2 - 75, this.height / 2 + 0, 150, 20,
				"Reload Config"));

		this.sliderSize = new GuiSlideControl(3, this.width / 2 - 75, this.height / 2 + 22, 150, 20,
				"Size: ", 1.0F, 100.0F, new Integer(RenderPlayerMod.size).floatValue(), true);
		this.sliderHorizontally = new GuiSlideControl(4, this.width / 2 - 75, this.height / 2 + 44, 150, 20,
				"Horizontally: ", -150.0F, 150.0F, RenderPlayerMod.horizontally, true);
		this.sliderVertically = new GuiSlideControl(5, this.width / 2 - 75, this.height / 2 + 66, 150, 20,
				"Vertically: ", -150.0F, 150.0F, RenderPlayerMod.vertically, true);
		this.buttonList.add(sliderSize);
		this.buttonList.add(sliderHorizontally);
		this.buttonList.add(sliderVertically);

		this.buttonList.add(this.buttonResetSize = new GuiButton(6, this.width / 2 + 77, this.height / 2 + 22, 20, 20, "-"));
		this.buttonList.add(this.buttonHorizontallyReset = new GuiButton(7, this.width / 2 + 77, this.height / 2 + 44, 20, 20, "-"));
		this.buttonList.add(this.buttonVerticallyReset = new GuiButton(8, this.width / 2 + 77, this.height / 2 + 66, 20, 20, "-"));

		this.buttonList.add(this.buttonOpenGithub = new GuiButton(9, this.width / 2 - 75, this.height / 2 + 88, 150, 20,
				"Open Github"));
	}

	public void display() {
		FMLCommonHandler.instance().bus().register(this);
	}

	@SubscribeEvent
	public void onClientTick(TickEvent.ClientTickEvent event) {
		FMLCommonHandler.instance().bus().unregister(this);
		Minecraft.getMinecraft().displayGuiScreen(this);
	}

	public void drawScreen(int x, int y, float partialTicks) {
		super.drawDefaultBackground();
		drawCenteredString(this.mc.fontRendererObj, "RenderPlayerMod 1.8.9", this.width / 2, 4, -1);
		super.drawScreen(x, y, partialTicks);
	}

	protected void mouseClicked(int x, int y, int time) {
		int minX = RenderPlayerMod.counterPosX - 25;
		int minY = RenderPlayerMod.counterPosY - 65;
		int maxX = RenderPlayerMod.counterPosX + 10;
		int maxY = RenderPlayerMod.counterPosY + 10;
		if ((x >= minX) && (x <= maxX) && (y >= minY) && (y <= maxY)) {
			this.isDragging = true;
			this.lastX = x;
			this.lastY = y;
		}
		try {
			super.mouseClicked(x, y, time);
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	protected void mouseReleased(int x, int y, int which) {
		if ((which == 0) && (this.isDragging)) {
			this.isDragging = false;
		}
		super.mouseReleased(x, y, which);
	}

	protected void mouseClickMove(int x, int y, int lastButtonClicked, long timeSinceClick) {
		if (this.isDragging) {
			RenderPlayerMod.counterPosX += x - this.lastX;
			RenderPlayerMod.counterPosY += y - this.lastY;
			this.lastX = x;
			this.lastY = y;
		}
		super.mouseClickMove(x, y, lastButtonClicked, timeSinceClick);
	}

	private String[] links = {"https://github.com/SimplyRin/RenderPlayerMod-1.8.9", "https://github.com/S4wa/RenderPlayerMod-1.8.9"};
	protected void actionPerformed(GuiButton button) {
		if (button == this.buttonToggle) {
			RenderPlayerMod.enabled = !RenderPlayerMod.enabled;
			this.buttonToggle.displayString = "Enabled: " + RenderPlayerMod.enabled;

		} else if (button == this.buttonReset) {
			RenderPlayerMod.counterPosX = 20;
			RenderPlayerMod.counterPosY = 60;

		} else if (button == this.buttonReload) {
			RenderPlayerMod.reloadSettings();

		} else if (button == this.sliderSize) {
			RenderPlayerMod.size = this.sliderSize.GetValueAsInt();

		} else if (button == this.buttonResetSize) {
			RenderPlayerMod.size = 30;
			this.sliderSize.displayString = "Size: " + 30;

		} else if (button == this.sliderHorizontally) {
			RenderPlayerMod.horizontally = this.sliderHorizontally.GetValueAsFloat();

		} else if (button == this.sliderVertically) {
			RenderPlayerMod.vertically = this.sliderVertically.GetValueAsFloat();

		} else if (button == this.buttonHorizontallyReset) {
			RenderPlayerMod.horizontally = 0.0F;
			this.sliderHorizontally.displayString = "Horizontally: " + 0;

		} else if (button == this.buttonVerticallyReset) {
			RenderPlayerMod.vertically = 0.0F;
			this.sliderVertically.displayString = "Vertically: " + 0;

		} else if (button == this.buttonOpenGithub) {
			for (String s : links) {
				try {
					Desktop.getDesktop().browse(new URI(s));
				} catch (IOException e) {
					e.printStackTrace();
				} catch (URISyntaxException e) {
					e.printStackTrace();
				}
			}
		}
	}

	public void onGuiClosed() {
		RenderPlayerMod.saveSettings();
	}

	public boolean doesGuiPauseGame() {
		return false;
	}
}
